filament/backend/src/metal/MetalHandles.mm

```diff
@@ -116,38 +119,60 @@ static inline MTLStorageMode getMetalStorageMode(TextureFormat format) {
 
 MetalIndexBuffer::~MetalIndexBuffer() {
     [buffer release];
-};
+}
 
-MetalUniformBuffer::MetalUniformBuffer(id<MTLDevice> device, size_t size) : HwUniformBuffer(),
-        size(size) {
+MetalUniformBuffer::MetalUniformBuffer(MetalContext& context, size_t size) : HwUniformBuffer(),
+        size(size), context(context) {
+    ASSERT_PRECONDITION(size > 0, "Cannot create Metal uniform with size %d.", size);
     // If the buffer is less than 4K in size, we don't use an explicit buffer and instead use
     // immediate command encoder methods like setVertexBytes:length:atIndex:.
     if (size <= 4 * 1024) {   // 4K
-        buffer = nil;
+        bufferPoolEntry = nullptr;
         cpuBuffer = malloc(size);
-    } else {
-        buffer = [device newBufferWithLength:size
-                                     options:MTLResourceStorageModeShared];
-        cpuBuffer = nullptr;
     }
 }
 
 MetalUniformBuffer::~MetalUniformBuffer() {
-    if (buffer) {
-        [buffer release];
-    } else if (cpuBuffer) {
+    if (cpuBuffer) {
         free(cpuBuffer);
     }
 }
 
 void MetalUniformBuffer::copyIntoBuffer(void* src, size_t size) {
+    if (size <= 0) {
+        return;
+    }
+    ASSERT_PRECONDITION(size <= this->size, "Attempting to copy %d bytes into a uniform of size %d",
+            size, this->size);
+
     // Either copy into the Metal buffer or into our cpu buffer.
-    if (buffer) {
-        memcpy(buffer.contents, src, size);
-    } else {
-        assert(cpuBuffer);
+    if (cpuBuffer) {
         memcpy(cpuBuffer, src, size);
+        return;
     }
+
+    bufferPoolEntry = context.bufferPool->acquireBuffer(size);
+    memcpy(static_cast<uint8_t*>(bufferPoolEntry->buffer.contents), src, size);
+
+    const MetalBufferPoolEntry* entry = this->bufferPoolEntry;
+    MetalBufferPool* pool = context.bufferPool;
+    // Important to copy the bufferPool and entry pointers into separate variables so that the block
+    // does not capture "this", which may not be valid by the time the frame has completed (if this
+    // uniform is destroyed, for example).
+    [context.currentCommandBuffer addCompletedHandler:^(id<MTLCommandBuffer> commandBuffer) {
+        pool->releaseBuffer(entry);
```

Comment:
I'm trying to understand ...  after the current command buffer finishes up, the low-level `MTLBuffer` will be slated for destruction 10 frames later.  But, if the higher-level `MetalUniformBuffer` is still in use 10 frames later, then won't its `getGpuBuffer` method return a non-null value pointing to garbage?

Reply:
I sort of get it: for Metal, we're using a pool to achieve double-buffering of a shared memory area, whereas for Vulkan we're using a pool as a staging area into GPU-only memory.  

However I still think the lifetime of the underlying metal buffers might be incorrect, or needs further explanation in the comment block.

Reply:
Yeah, it's for pooling the shared memory area. You're right though, there's a bug here. This implementation assumes that uniforms are updated at least once per 10 frames. `releaseBuffer` should only be called if the command buffer has completed **and** the `MetalUniformBuffer` no longer needs the buffer (if the uniform has been updated and a new buffer obtained or if the uniform has been destroyed).